python/cudf/cudf/core/buffer.py

```diff
@@ -89,6 +90,37 @@ def empty(cls, size):
         dbuf = DeviceBuffer(size=size)
         return Buffer(dbuf)
 
+    def serialize(self):
+        from distributed.protocol.cuda import cuda_dumps
```

Comment:
Do we want to explicitly make dask/distributed a dependency of cudf? I realize it is packaged as such, but so far we've avoided this I think.

Reply:
It's no more a dependency here than [it already is]( https://github.com/rapidsai/cudf/blob/b451ce60ae6310d9cb7b5a2d7fabc8dae79f581d/python/cudf/cudf/comm/serialize.py#L7-L8 ).

It's also worth noting this is a soft dependency though. If people try to run these methods, things won't work. However IIUC after looking at other cuDF classes, these methods are only meant to be used with Dask. Plus `Buffer`s still otherwise work without Dask.

Please let me know if you have more concerns I've missed.

Reply:
I think the concern here is that the serialize file is really only used by dask and not in the cudf codebase.  Where as including dask at the core of cudf would more directly require dask when serializing any cudf object.

Reply:
Yeah I'm not really sure how else to accomplish this, but would be open to suggestions 🙂

Reply:
I think we want to avoid this kind of entanglement as debugging between dask and cudf become challenging quite quickly.  Instead, perhaps we can ask how much of a priority it is to have a numba fallback. I suspect not vey high and we can rely on the rmm ptrs directly .  @kkraus14 do you have thoughts here ?

Reply:
It's probably also worth benchmarking to see what kind of a difference we see here. Was going to ask someone to try, but wanted to wait until we got CI sorted.

Reply:
>  I suspect not vey high and we can rely on the rmm ptrs directly . @kkraus14 do you have thoughts here ?

Rip out the numba, use RMM directly for sure.

Reply:
I also am not in favor of tying `serialize` and `deserialize` functions to Dask. These could be used by other systems or users directly who are not using Dask.

Reply:
Moving the code to the comms file seems like a good idea. Ideally serialize should get us to something that is as close to the physical memory as possible. Buffer is essentially that where we shouldn't need to define any additional serialization / deserialization code for it, but the distributed / dask integration should be outside of the class / functions.

Reply:
Yep, ideally I just want to hand the underlying data to Dask and say, "Hey, please figure out how to serialize this for me."

In the worst case we might still need some fallback to Numba, but I hope that is never needed. Maybe I can pick your brain and we can figure out how to handle those cases better too 😄

Well let me try moving the code and then we can see where things stand 🙂

Reply:
Do you know of any cases where we currently fall back to Numba? A Buffer should essentially look like a subset of Numba so I can't imagine where we'd want to fall back.

Reply:
Not totally sure. 😅 This is where I probably need guidance from you and Ashwin. 😉 

The case of [passing in a pointer address]( https://github.com/rapidsai/cudf/blob/a512e92e62ffe67ea8ab1d772c38ab43c5305da8/python/cudf/cudf/core/buffer.py#L43-L48 ) might be one (especially if there is no `owner` provided). We don't have a good way to handle [`_DevicePointer`s]( https://github.com/rapidsai/cudf/blob/a512e92e62ffe67ea8ab1d772c38ab43c5305da8/python/cudf/cudf/core/buffer.py#L32-L35 ) currently. Maybe there are other cases? Thoughts?

Reply:
You shouldn't need to reach into the owner attribute so whether it's a cupy array, numba array, DevicePointer, etc. shouldn't matter. Given we have a pointer and size for `Buffer` we have all of the information we need to send to UCX, no?

Reply:
Maybe, I'll think about this a bit more. Perhaps we should scrap this approach and try another one.